org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinInlineFilter.java

```diff
@@ -0,0 +1,116 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2018 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.LineNumberNode;
+import org.objectweb.asm.tree.MethodNode;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.StringReader;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Filters out instructions that were inlined by Kotlin compiler.
+ */
+public final class KotlinInlineFilter implements IFilter {
+
+	public void filter(final MethodNode methodNode,
+			final IFilterContext context, final IFilterOutput output) {
+		if (context.getSourceDebugExtension() == null) {
+			return;
+		}
+		if (!context.getClassAnnotations()
+				.contains(KotlinGeneratedFilter.KOTLIN_METADATA_DESC)) {
+			return;
+		}
+
+		final int max = parseSMAP(context.getSourceDebugExtension());
```

Comment:
Currently filters are stateless and do processing of one method after another, so the same SMAP will be parsed multiple times - need to think how to avoid this without introduction of coupling with Kotlin's SMAP outside of filter. Either allow filter to somehow store some information outside? Or simply creation of instances of filters for each analyzed class, so that they can keep per-class state?

Reply:
Another option would be to return an `SourceDebugExtension` object from `IFilterContext.getSourceDebugExtension()'. Such an instance would do the String parsing  lazily and only once.

Reply:
This is what I called 

> coupling with Kotlin's SMAP outside of filter

unless the parsing is generalized from Kotlin's SMAP to just SMAP, but then not even necessarily that `SourceDebugExtension` attribute contains data in JSR-45 format. JVM Spec §4.7.11 says

> The debug_extension array holds extended debugging information which has no semantic effect on the Java Virtual Machine. The information is represented using a modified UTF-8 string with no terminating zero byte.

And IMO that's why in ASM it is just a `String` and ASM doesn't parse it.

But maybe something like `tryParseAsSMAP` won't damage API clearness.

Reply:
As IMO a good first solution made filters stateful.

Reply:
Now method `Filters.all` creates new instances of all filters, however there is another option - keep all existing filters except new one as `static`. Something like

```
/**
 * Filter that combines other filters.
 */
public final class Filters implements IFilter {

	/**
	 * Filter that does nothing.
	 */
	public static final IFilter NONE = new Filters();

	/**
	 * Stateless filters.
	 */
	private static final IFilter STATELESS = new Filters(new EnumFilter(),
			new SyntheticFilter(), new SynchronizedFilter(),
			new TryWithResourcesJavac11Filter(),
			new TryWithResourcesJavacFilter(), new TryWithResourcesEcjFilter(),
			new FinallyFilter(), new PrivateEmptyNoArgConstructorFilter(),
			new StringSwitchJavacFilter(), new StringSwitchEcjFilter(),
			new EnumEmptyConstructorFilter(), new AnnotationGeneratedFilter(),
			new KotlinGeneratedFilter(), new KotlinLateinitFilter(),
			new KotlinWhenFilter(), new KotlinWhenStringFilter(),
			new KotlinUnsafeCastOperatorFilter(),
			new KotlinDefaultArgumentsFilter());

	private final IFilter[] filters;

	/**
	 * Creates filter that combines all other filters.
	 * 
	 * @return filter that combines all other filters
	 */
	public static IFilter all() {
		return new Filters(STATELESS, new KotlinInlineFilter());
	}
```

@marchof WDYT?

Reply:
@Godin Creating all filters per class isn't a big deal. Let's keep it simple for now.

Reply:
👍